File: packages/core-rust/src/webauthn.rs
use ciborium::Value;

use crate::authdata;
use crate::base64url;
use crate::clientdata;
use crate::composite;
use crate::cose;
use crate::es256;
use crate::mldsa65;
use crate::packed;
use crate::types::{
    AuthenticationInput, AuthenticationResult, RegistrationInput, RegistrationResult, WebAuthnError,
};

/// Helper to get a value from a COSE key CBOR map by integer key.
fn get_cose_int_value(map: &[(Value, Value)], key: i128) -> Option<i128> {
    map.iter()
        .find(|(k, _)| matches!(k, Value::Integer(i) if i128::from(*i) == key))
        .and_then(|(_, v)| match v {
            Value::Integer(i) => Some(i128::from(*i)),
            _ => None,
        })
}

fn get_cose_bytes(map: &[(Value, Value)], key: i128) -> Option<Vec<u8>> {
    map.iter()
        .find(|(k, _)| matches!(k, Value::Integer(i) if i128::from(*i) == key))
        .and_then(|(_, v)| match v {
            Value::Bytes(b) => Some(b.clone()),
            _ => None,
        })
}

/// Verify a WebAuthn registration ceremony.
pub fn verify_registration(input: RegistrationInput) -> Result<RegistrationResult, WebAuthnError> {
    // 1. Verify client data
    let client_data = clientdata::verify_client_data(
        &input.client_data_json,
        "webauthn.create",
        &input.expected_challenge,
        &input.expected_origin,
    )?;

    // 2. Decode attestation object (CBOR)
    let att_obj_bytes = base64url::decode(&input.attestation_object)?;
    let att_obj: Value = ciborium::from_reader(att_obj_bytes.as_slice())
        .map_err(|e| WebAuthnError::InvalidInput(format!("invalid attestation object CBOR: {}", e)))?;

    let att_map = match &att_obj {
        Value::Map(m) => m,
        _ => {
            return Err(WebAuthnError::InvalidInput(
                "attestation object not a map".into(),
            ))
        }
    };

    // Extract fmt
    let fmt = att_map
        .iter()
        .find(|(k, _)| matches!(k, Value::Text(s) if s == "fmt"))
        .and_then(|(_, v)| match v {
            Value::Text(s) => Some(s.as_str()),
            _ => None,
        })
        .unwrap_or("none");

    // Extract attStmt
    let att_stmt = att_map
        .iter()
        .find(|(k, _)| matches!(k, Value::Text(s) if s == "attStmt"))
        .and_then(|(_, v)| match v {
            Value::Map(m) => Some(m.as_slice()),
            _ => None,
        });

    // Extract authData
    let auth_data_bytes = att_map
        .iter()
        .find(|(k, _)| matches!(k, Value::Text(s) if s == "authData"))
        .and_then(|(_, v)| match v {
            Value::Bytes(b) => Some(b.clone()),
            _ => None,
        })
        .ok_or_else(|| WebAuthnError::InvalidInput("missing authData in attestation object".into()))?;

    // 3. Parse authenticator data
    let auth_data = authdata::parse_auth_data(&auth_data_bytes)?;

    // 4. Verify RP ID hash
    authdata::verify_rp_id_hash(&auth_data.rp_id_hash, &input.rp_id)?;

    // 5. Verify flags
    authdata::verify_flags(&auth_data, input.require_user_verification)?;

    // 6. Extract credential data
    let credential_id = auth_data
        .credential_id
        .as_ref()
        .ok_or_else(|| WebAuthnError::InvalidInput("no credential id in auth data".into()))?
        .clone();

    let cred_public_key_bytes = auth_data
        .credential_public_key
        .as_ref()
        .ok_or_else(|| WebAuthnError::InvalidInput("no public key in auth data".into()))?
        .clone();

    // 7. Verify attestation
    match fmt {
        "none" => {
            // No verification needed
        }
        "packed" => {
            let stmt = att_stmt.ok_or(WebAuthnError::InvalidAttestationStatement)?;
            packed::verify_packed_attestation(
                stmt,
                &auth_data_bytes,
                &client_data.raw_bytes,
                &cred_public_key_bytes,
            )?;
        }
        _ => {
            return Err(WebAuthnError::UnsupportedAttestationFormat);
        }
    }

    Ok(RegistrationResult {
        credential_id,
        public_key_cose: cred_public_key_bytes,
        sign_count: auth_data.sign_count,
        rp_id_hash: auth_data.rp_id_hash,
        backup_eligible: auth_data.backup_eligible(),
        backup_state: auth_data.backup_state(),
    })
}

/// Verify a WebAuthn authentication ceremony.
pub fn verify_authentication(
    input: AuthenticationInput,
) -> Result<AuthenticationResult, WebAuthnError> {
    // 1. Verify client data
    let client_data = clientdata::verify_client_data(
        &input.client_data_json,
        "webauthn.get",
        &input.expected_challenge,
        &input.expected_origin,
    )?;

    // 2. Decode authenticator data
    let auth_data_bytes = base64url::decode(&input.authenticator_data)?;
    let auth_data = authdata::parse_auth_data(&auth_data_bytes)?;

    // 3. Verify RP ID hash
    authdata::verify_rp_id_hash(&auth_data.rp_id_hash, &input.rp_id)?;

    // 4. Verify flags
    authdata::verify_flags(&auth_data, input.require_user_verification)?;

    // 5. Check sign count rollback
    if input.stored_sign_count > 0 && auth_data.sign_count <= input.stored_sign_count {
        return Err(WebAuthnError::SignCountRollback);
    }

    // 6. Decode stored public key and verify signature
    let stored_key_bytes = base64url::decode(&input.stored_public_key_cose)?;
    let signature_bytes = base64url::decode(&input.signature)?;

    let cose_key: Value = ciborium::from_reader(stored_key_bytes.as_slice())
        .map_err(|e| WebAuthnError::InvalidInput(format!("invalid COSE key CBOR: {}", e)))?;

    let map = match &cose_key {
        Value::Map(m) => m.as_slice(),
        _ => return Err(WebAuthnError::InvalidInput("COSE key not a map".into())),
    };

    let alg = get_cose_int_value(map, cose::COSE_KEY_ALG)
        .ok_or(WebAuthnError::UnsupportedCoseAlgorithm)?;

    match alg {
        cose::COSE_ALG_ES256 => {
            let x = get_cose_bytes(map, cose::COSE_KEY_X)
                .ok_or_else(|| WebAuthnError::InvalidInput("missing x in COSE key".into()))?;
            let y = get_cose_bytes(map, cose::COSE_KEY_Y)
                .ok_or_else(|| WebAuthnError::InvalidInput("missing y in COSE key".into()))?;

            es256::verify_es256(
                &x,
                &y,
                &auth_data_bytes,
                &client_data.raw_bytes,
                &signature_bytes,
            )?;
        }
        cose::COSE_ALG_MLDSA65 => {
            let pub_key = get_cose_bytes(map, cose::COSE_KEY_PUB)
                .ok_or_else(|| {
                    WebAuthnError::InvalidInput("missing pub in COSE key".into())
                })?;

            mldsa65::verify_mldsa65(
                &pub_key,
                &auth_data_bytes,
                &client_data.raw_bytes,
                &signature_bytes,
            )?;
        }
        cose::COSE_ALG_COMPOSITE_MLDSA65_ES256 => {
            let pub_key = get_cose_bytes(map, cose::COSE_KEY_PUB)
                .ok_or_else(|| {
                    WebAuthnError::InvalidInput("missing pub in COSE key".into())
                })?;

            composite::verify_composite(
                &pub_key,
                &auth_data_bytes,
                &client_data.raw_bytes,
                &signature_bytes,
            )?;
        }
        _ => {
            return Err(WebAuthnError::UnsupportedCoseAlgorithm);
        }
    }

    Ok(AuthenticationResult {
        sign_count: auth_data.sign_count,
        backup_eligible: auth_data.backup_eligible(),
        backup_state: auth_data.backup_state(),
    })
}

File: packages/core-py/open_passkey/webauthn.py
"""Main WebAuthn verification functions."""

import hashlib
from dataclasses import dataclass

import cbor2

from .base64url import base64url_decode, base64url_encode
from .clientdata import verify_client_data
from .authdata import parse_authenticator_data, verify_rp_id_hash
from .cose import COSE_ALG_ES256, COSE_ALG_MLDSA65, COSE_ALG_COMPOSITE_MLDSA65_ES256
from .es256 import verify_es256_signature
from .mldsa65 import verify_mldsa65_signature
from .composite import verify_composite_signature
from .packed import verify_packed_self_attestation, verify_packed_full_attestation
from .errors import (
    UserPresenceRequiredError,
    UserVerificationRequiredError,
    InvalidBackupStateError,
    UnsupportedAttestationFormatError,
    InvalidAttestationStatementError,
    UnsupportedAlgorithmError,
    SignCountRollbackError,
)


@dataclass
class RegistrationResult:
    credential_id: str  # base64url
    public_key_cose: str  # base64url
    sign_count: int
    rp_id_hash: str  # base64url
    flags: int
    backup_eligible: bool
    backup_state: bool
    attestation_format: str


@dataclass
class AuthenticationResult:
    sign_count: int
    flags: int
    backup_eligible: bool
    backup_state: bool


def verify_registration(
    rp_id: str,
    expected_challenge: str,
    expected_origin: str,
    client_data_json: str,
    attestation_object: str,
    require_user_verification: bool = False,
) -> RegistrationResult:
    """Verify a WebAuthn registration ceremony.

    All binary inputs are base64url-encoded strings (no padding).
    """
    # Verify client data
    client_data_json_raw = verify_client_data(
        client_data_json, "webauthn.create", expected_challenge, expected_origin
    )

    # Decode CBOR attestation object
    att_obj_bytes = base64url_decode(attestation_object)
    att_obj = cbor2.loads(att_obj_bytes)

    fmt = att_obj.get("fmt", att_obj.get(b"fmt", ""))
    auth_data = att_obj.get("authData", att_obj.get(b"authData", b""))
    att_stmt_raw = att_obj.get("attStmt", att_obj.get(b"attStmt", {}))

    # Ensure fmt is a string
    if isinstance(fmt, bytes):
        fmt = fmt.decode("utf-8")

    # Ensure auth_data is bytes
    auth_data = bytes(auth_data)

    att_stmt = None
    if fmt == "none":
        pass
    elif fmt == "packed":
        # Normalize att_stmt keys
        normalized = {}
        for k, v in att_stmt_raw.items():
            key = k.decode("utf-8") if isinstance(k, bytes) else k
            normalized[key] = v

        alg = normalized.get("alg")
        sig = normalized.get("sig")
        if alg is None or sig is None:
            raise InvalidAttestationStatementError("missing alg or sig")
        att_stmt = {
            "alg": alg,
            "sig": sig,
        }
        x5c = normalized.get("x5c")
        if x5c is not None:
            att_stmt["x5c"] = x5c
    else:
        raise UnsupportedAttestationFormatError(fmt)

    # Parse authenticator data
    parsed = parse_authenticator_data(auth_data, True)

    # Verify RP ID hash
    verify_rp_id_hash(parsed.rp_id_hash, rp_id)

    # Verify UP flag
    if (parsed.flags & 0x01) == 0:
        raise UserPresenceRequiredError()

    # Verify UV flag if required
    if require_user_verification and (parsed.flags & 0x04) == 0:
        raise UserVerificationRequiredError()

    # BS must be 0 if BE is 0
    if (parsed.flags & 0x08) == 0 and (parsed.flags & 0x10) != 0:
        raise InvalidBackupStateError()

    # Verify packed attestation
    if fmt == "packed" and att_stmt is not None:
        client_data_hash = hashlib.sha256(client_data_json_raw).digest()
        x5c = att_stmt.get("x5c")
        if x5c and len(x5c) > 0:
            verify_packed_full_attestation(att_stmt, auth_data, client_data_hash)
        else:
            verify_packed_self_attestation(
                parsed.credential_key, auth_data, client_data_json_raw, bytes(att_stmt["sig"])
            )

    return RegistrationResult(
        credential_id=base64url_encode(parsed.credential_id),
        public_key_cose=base64url_encode(parsed.credential_key),
        sign_count=parsed.sign_count,
        rp_id_hash=base64url_encode(parsed.rp_id_hash),
        flags=parsed.flags,
        backup_eligible=(parsed.flags & 0x08) != 0,
        backup_state=(parsed.flags & 0x10) != 0,
        attestation_format=fmt,
    )


def verify_authentication(
    rp_id: str,
    expected_challenge: str,
    expected_origin: str,
    stored_public_key_cose: str,
    stored_sign_count: int,
    client_data_json: str,
    authenticator_data: str,
    signature: str,
    require_user_verification: bool = False,
) -> AuthenticationResult:
    """Verify a WebAuthn authentication ceremony.

    All binary inputs are base64url-encoded strings (no padding).
    """
    # Verify client data
    client_data_json_raw = verify_client_data(
        client_data_json, "webauthn.get", expected_challenge, expected_origin
    )

    # Decode authenticator data
    auth_data_raw = base64url_decode(authenticator_data)
    parsed = parse_authenticator_data(auth_data_raw, False)

    # Verify RP ID hash
    verify_rp_id_hash(parsed.rp_id_hash, rp_id)

    # Verify UP flag
    if (parsed.flags & 0x01) == 0:
        raise UserPresenceRequiredError()

    # Verify UV flag if required
    if require_user_verification and (parsed.flags & 0x04) == 0:
        raise UserVerificationRequiredError()

    # BS must be 0 if BE is 0
    if (parsed.flags & 0x08) == 0 and (parsed.flags & 0x10) != 0:
        raise InvalidBackupStateError()

    client_data_hash = hashlib.sha256(client_data_json_raw).digest()
    sig_bytes = base64url_decode(signature)

    # Decode stored public key to determine algorithm
    stored_key_bytes = base64url_decode(stored_public_key_cose)
    cose_key = cbor2.loads(stored_key_bytes)
    alg = cose_key.get(3)

    if alg == COSE_ALG_ES256:
        verify_es256_signature(stored_key_bytes, auth_data_raw, client_data_hash, sig_bytes)
    elif alg == COSE_ALG_MLDSA65:
        verify_mldsa65_signature(stored_key_bytes, auth_data_raw, client_data_hash, sig_bytes)
    elif alg == COSE_ALG_COMPOSITE_MLDSA65_ES256:
        verify_composite_signature(stored_key_bytes, auth_data_raw, client_data_hash, sig_bytes)
    else:
        raise UnsupportedAlgorithmError()

    # Sign count rollback detection
    if stored_sign_count > 0 and parsed.sign_count <= stored_sign_count:
        raise SignCountRollbackError()

    return AuthenticationResult(
        sign_count=parsed.sign_count,
        flags=parsed.flags,
        backup_eligible=(parsed.flags & 0x08) != 0,
        backup_state=(parsed.flags & 0x10) != 0,
    )
